server/ai_live_video.go

```diff
@@ -72,10 +72,17 @@ func startTrickleSubscribe(ctx context.Context, url *url.URL, params aiRequestPa
 	go func() {
 		defer r.Close()
 		retryCount := 0
-		// TODO check whether stream is actually terminated
-		//      so we aren't just looping unnecessarily
-		for retryCount < 10 {
-			_, err := ffmpeg.Transcode3(&ffmpeg.TranscodeOptionsIn{
+		for {
+			streamExists, err := params.mediaMTXClient.StreamExists()
```

Comment:
Feels odd to be doing RPC out to MediaMTX here; a better way to do this might be to check whether the live-video process is still active. One way would be to check whether the control channel still exists, but there are probably other ways too.

https://github.com/livepeer/go-livepeer/blob/ff2d395275175d7804900ed0e3b98a4d6d3298f5/server/ai_live_video.go#L115

We probably do need to go through the clean up logic and ensure things are tight (so we properly close the control channel along with everything else)

Reply:
Good call, I've changed now to check `params.node.LivePipelines`. Looks good in my tests, segmentation stops because of the mediamtx stream existence check, which then runs the cleanup function which allows the output stream to stop:

```
E1206 14:02:12.305958   74963 rtmp2segment.go:49] request_id=b5b828e1 capability=Live video to video url=https://192.168.122.69:8935/ai/trickle/51216041-out outputRTMPURL=rtmp://rtmp.livepeer.com/live/b7ca-f3im-c6ya-iwzj stream=stream source_id=48adb45c-ac98-44e1-9cc8-5b2fb8144614 source_type=rtmpConn remote_addr=127.0.0.1 Stopping segmentation, input stream does not exist. in=rtmp://127.0.0.1/stream err=get stream failed with status code: 404 body: {"error":"connection not found"}
I1206 14:02:12.306024   74963 rtmp2segment.go:68] source_type=rtmpConn remote_addr=127.0.0.1 request_id=b5b828e1 capability=Live video to video url=https://192.168.122.69:8935/ai/trickle/51216041-out outputRTMPURL=rtmp://rtmp.livepeer.com/live/b7ca-f3im-c6ya-iwzj stream=stream source_id=48adb45c-ac98-44e1-9cc8-5b2fb8144614 sent completion signal, now waiting
2024/12/06 14:02:12 INFO Completion signal received. Closing current segment to trigger EOF.
2024/12/06 14:02:12 INFO Got completion signal
E1206 14:02:13.052105   74963 ai_live_video.go:106] remote_addr=127.0.0.1 request_id=b5b828e1 capability=Live video to video url=https://192.168.122.69:8935/ai/trickle/51216041-out outputRTMPURL=rtmp://rtmp.livepeer.com/live/b7ca-f3im-c6ya-iwzj stream=stream source_id=48adb45c-ac98-44e1-9cc8-5b2fb8144614 source_type=rtmpConn Stopping output rtmp stream, input stream does not exist. err=Invalid data found when processing input
```